src/gpuarray_reduction.c

```diff
@@ -0,0 +1,548 @@
+ * @param [in]  isReduced  Either NULL, or an array of booleans of the same
+ *                         size as the dimensionality of the source tensor.
+ *                         Axis k is reduced if isReduced[k] is non-zero,
+ *                         and is preserved otherwise.
+ * @return GA_NO_ERROR if the operation was successful, or a non-zero error
+ *         code otherwise.
+ */
+
+GPUARRAY_PUBLIC int GpuArray_maxandargmax(GpuArray*       dstMax,
+                                          GpuArray*       dstArgmax,
+                                          const GpuArray* src,
+                                          const int*      isReduced){
+	/**
+	 * Generate kernel source code
+	 */
+	
+	const char*  dstMaxType      = gpuarray_get_type(src->typecode) -> cluda_name;
+	const char*  dstArgmaxType   = gpuarray_get_type(GA_LONG)       -> cluda_name;
+	const char*  s               = genkernel_maxandargmax(src->nd,
+	                                                      isReduced,
+	                                                      dstMaxType,
+	                                                      dstArgmaxType);
+	if(!s){return GA_MEMORY_ERROR;}
+	
+	/* Compile it */
+	const int    ARG_TYPECODE[8] = {
+		GA_POINTER, /* src */
+		GA_POINTER, /* srcSteps */
+		GA_POINTER, /* srcSize */
+		GA_POINTER, /* numBlk */
+		GA_POINTER, /* dstMax */
+		GA_POINTER, /* dstMaxSteps */
+		GA_POINTER, /* dstArgmax */
+		GA_POINTER  /* dstArgmaxSteps */
+	};
+	const size_t l = strlen(s);
+	GpuKernel kernel;
+	GpuKernel_init(&kernel, 0, 1, &s, &l, "maxandargmax",
+	               8, ARG_TYPECODE, 0, (char**)0);
+	
+	/* Invoke it. */
+	invokeMaxAndArgmax(&kernel, src, isReduced);
+
+	/* Return error code */
+	return GA_NO_ERROR;
+}
+
+/**
+ * Count the number of dimensions to be reduced.
+ */
+
+static int  getRdxIdx(const int numIdx, const int* isReduced){
+	int i, countReduced;
+	for(i=0, countReduced = 0;i<numIdx;i++){
+		countReduced += !!isReduced[i];
+	}
+	return countReduced;
+}
+
+/**
+ * @brief Generate the kernel code for MaxAndArgmax.
+ * 
+ * @param [in]  numIdx
+ * @param [in]  isReduced
+ * @param [in]  dstMaxType
+ * @param [in]  dstArgmaxType
+ * @return A free()'able string containing source code implementing the
+ *         kernel, or else NULL.
+ */
+
+static char* genkernel_maxandargmax(const int       numIdx,
+                                    const int*      isReduced,
+                                    const char*     dstMaxType,
+                                    const char*     dstArgmaxType){
+	/* Obtain the parameters of the reduction. */
+	GEN_KERNEL_CTX ctx;
+	ctx.numIdx        = numIdx;
+	ctx.isReduced     = isReduced;
+	ctx.numRedIdx     = getRdxIdx(ctx.numIdx, ctx.isReduced);
+	ctx.numFreeIdx    = ctx.numIdx - ctx.numRedIdx;
+	ctx.dstMaxType    = dstMaxType;
+	ctx.dstArgmaxType = dstArgmaxType;
+	
+	/**
+	 * Allocate string buffer.
+	 */
+	
+	strb* s = strb_alloc(5*1024);
```

Comment:
@abergeron A bit off-topic, but if `STRB_STATIC_INIT` were defined as `(strb){NULL, 0, 0}` instead of `{NULL, 0, 0}`, then it could be used both at initialization time (`strb foo = STRB_STATIC_INIT;`) and later on (`foo = STRB_STATIC_INIT;`).